art/estimators/classification/pytorch.py

```diff
@@ -701,3 +702,117 @@ def get_layers(self) -> List[str]:
+            x, y = defence(x, y)
+        else:
+            # Check if all defences are implemented in PyTorch.
+            for defence in self.preprocessing_defences:
+                if not hasattr(defence, "forward"):
+                    raise NotImplementedError(f"{defence.__class__} doesn't have method forward().")
+
+            # Convert np arrays to torch tensors.
+            x = torch.tensor(x, device=self._device)
+            if y is not None:
+                y = torch.tensor(y, device=self._device)
+
+            with torch.no_grad():
+                for defence in self.preprocessing_defences:
+                    if fit:
+                        if defence.apply_fit:
+                            x, y = defence.forward(x, y)
+                    else:
+                        if defence.apply_predict:
+                            x, y = defence.forward(x, y)
+
+            # Convert torch tensors back to np arrays.
+            x = x.cpu().numpy()
+            if y is not None:
+                y = y.cpu().numpy()
+
+        return x, y
+
+    def _apply_preprocessing_defences_gradient(self, x, gradients, fit=False):
+        """
+        Apply the backward pass to the gradients through all preprocessing defences that have been applied to `x`
+        and `y` in the forward pass. This function is should only be called from function
+        `_apply_preprocessing_gradient`.
+
+        The method overrides art.estimators.estimator::LossGradientsMixin._apply_preprocessing_defences_gradient().
+        It requires all defenses to have a method estimate_forward().
+        It converts numpy arrays to PyTorch tensors first, then chains a series of defenses by calling
+        defence.estimate_forward() which contains differentiable estimate of the operations. At the end,
+        it converts PyTorch tensors back to numpy arrays.
+
+        :param x: Samples.
+        :type x: Format as expected by the `model`
+        :param gradients: Gradients before backward pass through preprocessing defences.
+        :type gradients: Format as expected by the `model`
+        :param fit: `True` if the gradients are computed during training.
+        :return: Gradients after backward pass through preprocessing defences.
+        :rtype: Format as expected by the `model`
+        """
+        if not hasattr(self, "preprocessing_defences") or \
+           self.preprocessing_defences is None or \
+           len(self.preprocessing_defences) == 0:
+            return gradients
+
+        if len(self.preprocessing_defences) == 1:
+            # Compatible with non-PyTorch defences if no chaining.
+            defence = self.preprocessing_defences[0]
+            gradients = defence.estimate_gradient(x, gradients)
+        else:
+            # Check if all defences are implemented in PyTorch.
+            for defence in self.preprocessing_defences:
+                if not hasattr(defence, "estimate_forward"):
```

Comment:
After introducing the new abstract base class `PreprocessorPyTorch` above we can use to test for the correct type here.